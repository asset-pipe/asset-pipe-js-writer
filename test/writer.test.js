'use strict';

const getStream = require('./helpers/get-stream');
const Writer = require('../');
const prettier = require('prettier');

const clean = body => prettier.format(JSON.stringify(body, null, 2));

test('writer produces a feed of dependencies with ids hashed', async () => {
    expect.hasAssertions();
    const writer = new Writer('./test/mock/main.js');
    const result = await getStream(writer.bundle());

    expect(clean(result)).toMatchSnapshot();
});

test('options object passes options on to browserify', async () => {
    expect.hasAssertions();
    const writer = new Writer(['./test/mock/main.js'], {
        debug: true,
    });
    const result = await getStream(writer.bundle());

    expect(clean(result)).toMatchSnapshot();
});

test('bundle option allows getting a bundle instead of a feed', async () => {
    expect.hasAssertions();
    const writer = new Writer('./test/mock/main.js', {}, true);
    const result = await getStream(writer.bundle());

    expect(result).toMatchSnapshot();
});

test('module throws if `files` argument is not an array', async () => {
    expect.hasAssertions();
    const run = () => new Writer(null, {}, true);
    expect(run).toThrowErrorMatchingSnapshot();
});

test('feed is not deduped', async () => {
    expect.hasAssertions();
    const writer = new Writer('./test/mock/no-dedupe.js', {});
    const result = await getStream(writer.bundle());

    expect(clean(result)).toMatchSnapshot();
});

test('different modules with identical source code do produce same id hash.', async () => {
    expect.hasAssertions();
    const writer = new Writer('./test/mock/identical-implementation');
    const feed = await getStream(writer.bundle());

    let identicalHashes = false;
    const hashes = new Map();
    for (const feedItem of feed) {
        if (hashes.has(feedItem.id)) {
            identicalHashes = true;
            break;
        }
        hashes.set(feedItem.id, true);
    }
    expect(identicalHashes).toBe(false);
});
