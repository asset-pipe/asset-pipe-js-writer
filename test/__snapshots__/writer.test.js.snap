// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle option allows getting a bundle instead of a feed 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

var fizzbuzz = require('asset-pipe-test-es5b');
var util = require('./util.js');

console.log(util.helloWorld());
console.log(util.helloNumber());
fizzbuzz({s:4, e:20}, function (mgs) {
    console.log(mgs);
});

},{\\"./util.js\\":2,\\"asset-pipe-test-es5b\\":3}],2:[function(require,module,exports){
'use strict';

var isPrime = require('asset-pipe-test-es5c');

function hello() {
    return 'Hello world!';
}

function numb(num) {
    if (isPrime(num)) {
        return 'Not a prime number';
    }
    return 'A prime number';
}

module.exports.helloWorld = hello;
module.exports.helloNumber = numb;

},{\\"asset-pipe-test-es5c\\":4}],3:[function(require,module,exports){
'use strict';

var isPrime = require('asset-pipe-test-es5c');

/**
 * Run through FizzBuzz for n integers
 *
 * @param {Number} config.s The number to start at
 * @param {Number} config.e The number to end at
 * @param {function} callback
 */

function fizzbuzz (config, callback) {
    for (var i = config.s; i <= config.e; i++) {
        var prime = '';
        if (config.p && isPrime(i)) {
            prime = ' Prime';
        }

        if (!(i % 15)) {
            callback('FizzBuzz');
            continue;
        } else if (!(i % 3)) {
            callback('Fizz' + prime);
            continue;
        } else if (!(i % 5)) {
            callback('Buzz' + prime);
            continue;
        }
        callback(prime.trim() || i);
    }
};

module.exports = fizzbuzz;

},{\\"asset-pipe-test-es5c\\":4}],4:[function(require,module,exports){
'use strict';

 function prime (num) {
    if (num < 4) {
        return true;
    }
    var max = Math.ceil(Math.sqrt(num));
    for (var i = 2; i <= max; i++) {
        if (num % i === 0) {
            return false;
        }
    }
    return true;
};

module.exports = prime;

},{}],5:[function(require,module,exports){
'use strict';

require('asset-pipe-test-es5a');
console.log('main says hello');

},{\\"asset-pipe-test-es5a\\":1}]},{},[5]);
"
`;

exports[`feed is not deduped 1`] = `
"[
  {
    id: \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\",
    source:
      \\"'use strict';\\\\n\\\\nvar fizzbuzz = require('asset-pipe-test-es5b');\\\\nvar util = require('./util.js');\\\\n\\\\nconsole.log(util.helloWorld());\\\\nconsole.log(util.helloNumber());\\\\nfizzbuzz({s:4, e:20}, function (mgs) {\\\\n    console.log(mgs);\\\\n});\\\\n\\",
    deps: {
      \\"./util.js\\":
        \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
      \\"asset-pipe-test-es5b\\":
        \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/main.js\\",
    index: 1,
    indexDeps: {
      \\"./util.js\\": 2,
      \\"asset-pipe-test-es5b\\": 3
    }
  },
  {
    id: \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\nfunction hello() {\\\\n    return 'Hello world!';\\\\n}\\\\n\\\\nfunction numb(num) {\\\\n    if (isPrime(num)) {\\\\n        return 'Not a prime number';\\\\n    }\\\\n    return 'A prime number';\\\\n}\\\\n\\\\nmodule.exports.helloWorld = hello;\\\\nmodule.exports.helloNumber = numb;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/util.js\\",
    index: 2,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    }
  },
  {
    id: \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\n/**\\\\n * Run through FizzBuzz for n integers\\\\n *\\\\n * @param {Number} config.s The number to start at\\\\n * @param {Number} config.e The number to end at\\\\n * @param {function} callback\\\\n */\\\\n\\\\nfunction fizzbuzz (config, callback) {\\\\n    for (var i = config.s; i <= config.e; i++) {\\\\n        var prime = '';\\\\n        if (config.p && isPrime(i)) {\\\\n            prime = ' Prime';\\\\n        }\\\\n\\\\n        if (!(i % 15)) {\\\\n            callback('FizzBuzz');\\\\n            continue;\\\\n        } else if (!(i % 3)) {\\\\n            callback('Fizz' + prime);\\\\n            continue;\\\\n        } else if (!(i % 5)) {\\\\n            callback('Buzz' + prime);\\\\n            continue;\\\\n        }\\\\n        callback(prime.trim() || i);\\\\n    }\\\\n};\\\\n\\\\nmodule.exports = fizzbuzz;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5b/main.js\\",
    index: 3,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    }
  },
  {
    id: \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\",
    source:
      \\"'use strict';\\\\n\\\\n function prime (num) {\\\\n    if (num < 4) {\\\\n        return true;\\\\n    }\\\\n    var max = Math.ceil(Math.sqrt(num));\\\\n    for (var i = 2; i <= max; i++) {\\\\n        if (num % i === 0) {\\\\n            return false;\\\\n        }\\\\n    }\\\\n    return true;\\\\n};\\\\n\\\\nmodule.exports = prime;\\\\n\\",
    deps: {},
    file: \\"node_modules/asset-pipe-test-es5c/main.js\\",
    index: 4,
    indexDeps: {}
  },
  {
    id: \\"17cec02a29ba2621285a0c6928e7b6f91f9e10cbaa45be1eddbf4faed55c8b3d\\",
    source: \\"'use strict';\\\\n\\\\nrequire('./main');\\\\n\\",
    deps: {
      \\"./main\\":
        \\"6bc16f4686fe6cf0bb03134e009a29a452568f28f6b0e1174f8d75d47366a6ec\\"
    },
    file: \\"test/mock/common.js\\",
    index: 5,
    indexDeps: {
      \\"./main\\": 6
    }
  },
  {
    id: \\"6bc16f4686fe6cf0bb03134e009a29a452568f28f6b0e1174f8d75d47366a6ec\\",
    source:
      \\"'use strict';\\\\n\\\\nrequire('asset-pipe-test-es5a');\\\\nconsole.log('main says hello');\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5a\\":
        \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\"
    },
    file: \\"test/mock/main.js\\",
    index: 6,
    indexDeps: {
      \\"asset-pipe-test-es5a\\": 1
    }
  },
  {
    entry: true,
    expose: false,
    file: \\"test/mock/no-dedupe.js\\",
    id: \\"c98cd71bd49ced468ea7011b8733d1a4f940f28f9336797fe7c24687d2d82dc6\\",
    order: 0,
    source: \\"'use strict';\\\\n\\\\nrequire('./common');\\\\nrequire('./main');\\\\n\\",
    deps: {
      \\"./common\\":
        \\"17cec02a29ba2621285a0c6928e7b6f91f9e10cbaa45be1eddbf4faed55c8b3d\\",
      \\"./main\\":
        \\"6bc16f4686fe6cf0bb03134e009a29a452568f28f6b0e1174f8d75d47366a6ec\\"
    },
    index: 7,
    indexDeps: {
      \\"./common\\": 5,
      \\"./main\\": 6
    }
  }
];
"
`;

exports[`module throws if \`files\` argument is not an array 1`] = `"Expected 'files' argument to @asset-pipe/js-writer constructor to be either a string or an array. Instead got 'object'"`;

exports[`options object passes options on to browserify 1`] = `
"[
  {
    id: \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\",
    source:
      \\"'use strict';\\\\n\\\\nvar fizzbuzz = require('asset-pipe-test-es5b');\\\\nvar util = require('./util.js');\\\\n\\\\nconsole.log(util.helloWorld());\\\\nconsole.log(util.helloNumber());\\\\nfizzbuzz({s:4, e:20}, function (mgs) {\\\\n    console.log(mgs);\\\\n});\\\\n\\",
    deps: {
      \\"./util.js\\":
        \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
      \\"asset-pipe-test-es5b\\":
        \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/main.js\\",
    index: 1,
    indexDeps: {
      \\"./util.js\\": 2,
      \\"asset-pipe-test-es5b\\": 3
    },
    sourceRoot: \\"file://localhost\\",
    sourceFile: \\"node_modules/asset-pipe-test-es5a/lib/main.js\\"
  },
  {
    id: \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\nfunction hello() {\\\\n    return 'Hello world!';\\\\n}\\\\n\\\\nfunction numb(num) {\\\\n    if (isPrime(num)) {\\\\n        return 'Not a prime number';\\\\n    }\\\\n    return 'A prime number';\\\\n}\\\\n\\\\nmodule.exports.helloWorld = hello;\\\\nmodule.exports.helloNumber = numb;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/util.js\\",
    index: 2,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    },
    sourceRoot: \\"file://localhost\\",
    sourceFile: \\"node_modules/asset-pipe-test-es5a/lib/util.js\\"
  },
  {
    id: \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\n/**\\\\n * Run through FizzBuzz for n integers\\\\n *\\\\n * @param {Number} config.s The number to start at\\\\n * @param {Number} config.e The number to end at\\\\n * @param {function} callback\\\\n */\\\\n\\\\nfunction fizzbuzz (config, callback) {\\\\n    for (var i = config.s; i <= config.e; i++) {\\\\n        var prime = '';\\\\n        if (config.p && isPrime(i)) {\\\\n            prime = ' Prime';\\\\n        }\\\\n\\\\n        if (!(i % 15)) {\\\\n            callback('FizzBuzz');\\\\n            continue;\\\\n        } else if (!(i % 3)) {\\\\n            callback('Fizz' + prime);\\\\n            continue;\\\\n        } else if (!(i % 5)) {\\\\n            callback('Buzz' + prime);\\\\n            continue;\\\\n        }\\\\n        callback(prime.trim() || i);\\\\n    }\\\\n};\\\\n\\\\nmodule.exports = fizzbuzz;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5b/main.js\\",
    index: 3,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    },
    sourceRoot: \\"file://localhost\\",
    sourceFile: \\"node_modules/asset-pipe-test-es5b/main.js\\"
  },
  {
    id: \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\",
    source:
      \\"'use strict';\\\\n\\\\n function prime (num) {\\\\n    if (num < 4) {\\\\n        return true;\\\\n    }\\\\n    var max = Math.ceil(Math.sqrt(num));\\\\n    for (var i = 2; i <= max; i++) {\\\\n        if (num % i === 0) {\\\\n            return false;\\\\n        }\\\\n    }\\\\n    return true;\\\\n};\\\\n\\\\nmodule.exports = prime;\\\\n\\",
    deps: {},
    file: \\"node_modules/asset-pipe-test-es5c/main.js\\",
    index: 4,
    indexDeps: {},
    sourceRoot: \\"file://localhost\\",
    sourceFile: \\"node_modules/asset-pipe-test-es5c/main.js\\"
  },
  {
    entry: true,
    expose: false,
    file: \\"test/mock/main.js\\",
    id: \\"6bc16f4686fe6cf0bb03134e009a29a452568f28f6b0e1174f8d75d47366a6ec\\",
    order: 0,
    source:
      \\"'use strict';\\\\n\\\\nrequire('asset-pipe-test-es5a');\\\\nconsole.log('main says hello');\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5a\\":
        \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\"
    },
    index: 5,
    indexDeps: {
      \\"asset-pipe-test-es5a\\": 1
    },
    sourceRoot: \\"file://localhost\\",
    sourceFile: \\"test/mock/main.js\\"
  }
];
"
`;

exports[`writer produces a feed of dependencies with ids hashed 1`] = `
"[
  {
    id: \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\",
    source:
      \\"'use strict';\\\\n\\\\nvar fizzbuzz = require('asset-pipe-test-es5b');\\\\nvar util = require('./util.js');\\\\n\\\\nconsole.log(util.helloWorld());\\\\nconsole.log(util.helloNumber());\\\\nfizzbuzz({s:4, e:20}, function (mgs) {\\\\n    console.log(mgs);\\\\n});\\\\n\\",
    deps: {
      \\"./util.js\\":
        \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
      \\"asset-pipe-test-es5b\\":
        \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/main.js\\",
    index: 1,
    indexDeps: {
      \\"./util.js\\": 2,
      \\"asset-pipe-test-es5b\\": 3
    }
  },
  {
    id: \\"d286af219ef4acea1b486c460bdc65443fba40bd913457ce3e256ccc221f2d07\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\nfunction hello() {\\\\n    return 'Hello world!';\\\\n}\\\\n\\\\nfunction numb(num) {\\\\n    if (isPrime(num)) {\\\\n        return 'Not a prime number';\\\\n    }\\\\n    return 'A prime number';\\\\n}\\\\n\\\\nmodule.exports.helloWorld = hello;\\\\nmodule.exports.helloNumber = numb;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5a/lib/util.js\\",
    index: 2,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    }
  },
  {
    id: \\"ecd571b921d0e82abe77bd7cbe5c97c9186504c86f840fb8bdc0a372310231c3\\",
    source:
      \\"'use strict';\\\\n\\\\nvar isPrime = require('asset-pipe-test-es5c');\\\\n\\\\n/**\\\\n * Run through FizzBuzz for n integers\\\\n *\\\\n * @param {Number} config.s The number to start at\\\\n * @param {Number} config.e The number to end at\\\\n * @param {function} callback\\\\n */\\\\n\\\\nfunction fizzbuzz (config, callback) {\\\\n    for (var i = config.s; i <= config.e; i++) {\\\\n        var prime = '';\\\\n        if (config.p && isPrime(i)) {\\\\n            prime = ' Prime';\\\\n        }\\\\n\\\\n        if (!(i % 15)) {\\\\n            callback('FizzBuzz');\\\\n            continue;\\\\n        } else if (!(i % 3)) {\\\\n            callback('Fizz' + prime);\\\\n            continue;\\\\n        } else if (!(i % 5)) {\\\\n            callback('Buzz' + prime);\\\\n            continue;\\\\n        }\\\\n        callback(prime.trim() || i);\\\\n    }\\\\n};\\\\n\\\\nmodule.exports = fizzbuzz;\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5c\\":
        \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\"
    },
    file: \\"node_modules/asset-pipe-test-es5b/main.js\\",
    index: 3,
    indexDeps: {
      \\"asset-pipe-test-es5c\\": 4
    }
  },
  {
    id: \\"a8307dbbd8da4b0752187d0f5aaf7184da3b85743e87ec77110ed88e36370047\\",
    source:
      \\"'use strict';\\\\n\\\\n function prime (num) {\\\\n    if (num < 4) {\\\\n        return true;\\\\n    }\\\\n    var max = Math.ceil(Math.sqrt(num));\\\\n    for (var i = 2; i <= max; i++) {\\\\n        if (num % i === 0) {\\\\n            return false;\\\\n        }\\\\n    }\\\\n    return true;\\\\n};\\\\n\\\\nmodule.exports = prime;\\\\n\\",
    deps: {},
    file: \\"node_modules/asset-pipe-test-es5c/main.js\\",
    index: 4,
    indexDeps: {}
  },
  {
    entry: true,
    expose: false,
    file: \\"test/mock/main.js\\",
    id: \\"6bc16f4686fe6cf0bb03134e009a29a452568f28f6b0e1174f8d75d47366a6ec\\",
    order: 0,
    source:
      \\"'use strict';\\\\n\\\\nrequire('asset-pipe-test-es5a');\\\\nconsole.log('main says hello');\\\\n\\",
    deps: {
      \\"asset-pipe-test-es5a\\":
        \\"0e0c593f5f89dd62af488699574c678b4726ad1e3de2e0a81141f3dfc02b6037\\"
    },
    index: 5,
    indexDeps: {
      \\"asset-pipe-test-es5a\\": 1
    }
  }
];
"
`;
